File: Code/Javascript/Factory/factory.js
function Train(id, color) {
    this.id = id;
    this.color = color;
}

function ModernTrain(id, color){
    Train.call(this, id, color)
}

ModernTrain.prototype = Object.create(Train.prototype);

function ClassicTrain(id, color){
    Train.call(this, id, color)
}
ClassicTrain.prototype = Object.create(Train);

function TrainCreator() { }

TrainCreator.prototype.createTrain = function() { throw Error('Not implemented') }

function ClassicTrainCreator(){}
ClassicTrainCreator.prototype = Object.create(TrainCreator.prototype);
ClassicTrainCreator.prototype.createTrain = () => new ClassicTrain(2, 'red');


function ModernTrainCreator(){}
ModernTrainCreator.prototype = Object.create(TrainCreator.prototype);
ModernTrainCreator.prototype.createTrain = () => new ModernTrain(1, 'white');

console.log(new ClassicTrainCreator().createTrain())
console.log(new ModernTrainCreator().createTrain())

File: Code/Go/Factory/factory.go
package main

import (
	"fmt"
)

type Train interface {
	getId() int
	getColor() string
}

type TrainStruct struct {
	id    int
	color string
}

func (t TrainStruct) getId() int {
	return t.id
}

func (t TrainStruct) getColor() string {
	return t.color
}

type ModernTrain struct {
	TrainStruct
}

type ClassicTrain struct {
	TrainStruct
}

type TrainCreator interface {
	createTrain() Train
}

type ModernTrainCreator struct {
	TrainCreator
}

type ClassicTrainCreator struct {
	TrainCreator
}

func (c ModernTrainCreator) createTrain() Train {
	return ModernTrain{TrainStruct{1, "white"}}
}

func (c ClassicTrainCreator) createTrain() Train {
	return ClassicTrain{TrainStruct{2, "red"}}
}

func describe(t Train) {
	fmt.Println("I am a train with id =", t.getId(), "and color = "+t.getColor())
}

func main() {
	describe(ModernTrainCreator{}.createTrain())
	describe(ClassicTrainCreator{}.createTrain())
}
